plugins/pulp_rpm/plugins/importers/yum/modularity.py

```diff
@@ -0,0 +1,284 @@
+    :param model: The ModulemdDefaults model object to add.
+    :type model: pulp_rpm.plugins.db.models.ModulemdDefaults
+    """
+    path = os.path.join(
+        get_working_directory(),
+        str(uuid4()))
+    document = default.dumps()
+    with open(path, 'w+') as fp:
+        fp.write(document)
+    model.checksum = sha256(document).hexdigest()
+    model.repo_id = repository.repo_id
+    try:
+        model.save_and_import_content(path)
+    except NotUniqueError:
+        updated = ModulemdDefaults.objects.get(**model.unit_key)
+        updated.update(
+            stream=model.stream,
+            profiles=model.profiles,
+            checksum=model.checksum)
+        updated.safe_import_content(path)
+        model = updated
+    repository_controller.associate_single_unit(repository, model)
+
+
+def add_defaults(repository, defaults):
+    """
+    Add the collection of modulemd-defaults content to the repository.
+
+    The content is created as needed and added to the repository.
+    Matching content already contained in the repository is compared
+    by checksum and updated as needed.
+
+    :param repository: A repository.
+    :type repository: pulp.server.db.model.Repository
+    :param defaults: A list of gi.repository.Modulemd.Module.
+    :type defaults: collections.Iterable
+    :return: The set of content unit keys contained in the repository
+             that are not contained in the collection of defaults to be added.
+    :rtype: set
+    """
+    wanted = set()
+    inventory = get_inventory(repository, Model=ModulemdDefaults)
+    for default in defaults:
+        model = modules.process_defaults_document(default)
+        model.repo_id = repository.repo_id
+        key = model.NAMED_TUPLE(**model.unit_key)
+        wanted.add(key)
+        try:
+            checksum = inventory[key]
+        except KeyError:
+            pass
+        else:
+            document = default.dumps()
+            model.checksum = sha256(document).hexdigest()
+            if checksum == model.checksum:
+                continue
+        add_default(repository, default, model)
+    remainder = set(inventory.iterkeys()).difference(wanted)
+    return remainder
+
+
+def remove_defaults(repository, defaults):
+    """
+    Remove ModulemdDefaults content from the repository.
+
+    :param repository: A repository.
+    :type repository: pulp.server.db.model.Repository
+    :param defaults: A set of ModulemdDefaults unit keys to be removed.
+    :type defaults: set
+    """
+    if not defaults:
+        return
+
+    uq = Q()
+    for key in defaults:
+        uq |= Q(**key._asdict())
+    aq = Q(unit_type_id=ModulemdDefaults.TYPE_ID)
+    q_set = repository_controller.find_repo_content_units(
+        repository,
+        repo_content_unit_q=aq,
+        yield_content_unit=True,
+        units_q=uq)
+
+    repository_controller.disassociate_units(repository, q_set)
+
+
+def load(metadata):
```

Comment:
I want to reuse this code in my migration which finds all the modulemd files and creates units out of them. The code below assumes that get_working_directory can be called, but I can't do it during the migration because it is not a task. Would it be possible to refactor this code so the temporary path could be provided as an argument to load() ?

Reply:
Sure thing.